ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -180,233 +203,113 @@ forkSyncStateOnTipPointChange registry menv =
-
-              -- The mempool capacity has been reached.
-              _ -> retry
-
-    validateNewUntilMempoolFull
-      :: [GenTx blk]
-      -- ^ The new transactions to validate.
-      -> ValidationResult blk
-      -- ^ The 'ValidationResult' from which to begin validating.
-      -> STM m (ValidationResult blk, [GenTx blk])
-      -- ^ The last 'ValidationResult' along with the remaining transactions
-      -- (those not yet validated due to the mempool capacity being reached).
-    validateNewUntilMempoolFull []        vr = pure (vr, [])
-    validateNewUntilMempoolFull (tx:txs') vr = do
-      -- Get the current mempool size
-      MempoolSize { msNumBytes = curSizeInBytes } <- getMempoolSize mpEnv
-
-      -- Determine what the mempool size would be if we were to commit the new
-      -- transactions we've validated thus far and also the next transaction
-      -- to validate, 'tx'.
-      -- If this value is greater than the 'MempoolCapacityBytes', then we
-      -- know not to continue validating at this time.
-      let newTxsBytes    = sum (txSize <$> vrNewValid vr) + txSize tx
-          newSizeInBytes = curSizeInBytes + newTxsBytes
-
-      -- The size of a mempool should never be greater than its capacity.
-      assert (curSizeInBytes <= mempoolCap) $
-        -- Here, we check whether we're at the mempool's capacity /before/
-        -- attempting to validate the next transaction.
-        if newSizeInBytes <= mempoolCap
-          then validateNewUntilMempoolFull txs' (extendVRNew mpEnvLedgerCfg tx vr)
-          else pure (vr, tx:txs') -- if we're at mempool capacity, we return the
-                                  -- last 'ValidationResult' as well as the
-                                  -- remaining transactions (those not yet
-                                  -- validated).
+    done acc toAdd = return (reverse acc, toAdd)
+
+    go acc []                     = done acc []
+    go acc toAdd@(firstTx:toAdd') = do
+      let firstTxSize = txSize firstTx
+      -- Note: we execute the continuation returned by 'atomically'
+      join $ atomically $ do
+        is <- readTVar mpEnvStateVar
+        let curSize = msNumBytes $ isMempoolSize is
+        if curSize + firstTxSize > capacity then
+          -- No space in the Mempool
+          return $ done acc toAdd
+        else do
+          let vr  = extendVRNew cfg firstTx $ validationResultFromIS is
+              is' = internalStateFromVR vr
+          unless (null (vrNewValid vr)) $
+            -- Each time we have found a valid transaction, we update the
+            -- Mempool. This keeps our STM transactions short, avoiding
+            -- repeated work.
+            writeTVar mpEnvStateVar is'
+
+          -- We only extended the ValidationResult with a single transaction
+          -- ('firstTx'). So if it's not in 'vrInvalid', it must be in
+          -- 'vrNewValid'.
+          return $ case listToMaybe (vrInvalid vr) of
```

Comment:
Do you think it'd be unnecessary to also `assert` that we somehow didn't get to this point with a `vrInvalid vr` whose length is greater than 1? 

Maybe I'm being overly cautious without good reason because this should be impossible given that you're only validating the first transaction at this point.

Reply:
Sure, I'll add it.